### DISCUSSION TITLE
Spec RFC

### DISCUSSION THREAD

**Discussion by 4kimov** [sqids/sqids-spec#2] [General]
The [spec](https://github.com/sqids/sqids-spec) is complete as far my todos.

The next steps would be to implement individual ports/repos and any change to the spec would have to be propagated to the other repos, so the best time to improve it is now.

The tests should show you how to use the spec. **Feel free to delete [tests/uniques.test.ts](tests/uniques.test.ts) while you're running this locally, because this file will take a while; otherwise, tests finish instantly.**

I'd appreciate feedback on anything, but especially:

- Whether the shuffle function produces the best possible shuffle (performance vs randomization)
- Optimizing the [default blacklist](https://github.com/sqids/sqids-blacklist). Who speaks multiple languages?
- Performance. Can you see any loops that can be optimized? (The spec repo is optimized for readability/clarity, so optimizations would have to be about the [Sqids algorithm](https://github.com/sqids/sqids-spec#-how-it-works), not JS/TS specific)
- Is there any better way to structure `isBlockedId()` function? There's a balance between matching too many words and not matching enough.

If no major issues, we can start implementing individual repos in about a week.

**Comment by tzvetkoff** [comment]
OK, I re-reviewed the spec and I have an issue with the default blacklist.

Having a default blacklist in the library is a nice thing. Forcing it on each implementation is not.
That been said, I'd like to have the _option_ to initialize it with an empty blacklist. Current spec always prepends the default blacklist, which is something I'd rather not do.

Other than that, current blacklist search logic is just O(scary).

**Reply by 4kimov** [reply]
Thank you for taking a look.

> That been said, I'd like to have the option to initialize it with an empty blacklist.

Initially I was looking at a way to give the user a constructor array param so they can opt into different languages manually, eg:

```javascript
const sqids = new Sqids({
  ...defaultOptions,
  blacklist: [...EN, ...ES, ...IT, "custom1", "custom2"],
});
```

Pros: more fine-tuned control + faster encoding *if blank blacklist*, Cons: more configs to deal with for the end-user + empty blacklist by default(?). I do enjoy the simplicity of not having to think about it as an end-user, so I did default to auto opt in; however, I'm not married to the idea and would love to hear what others expect. What's a better way to design this?

> Other than that, current blacklist search logic is just O(scary).

How do we make it O(not-so-scary)?

**Reply by tzvetkoff** [reply]
I'd really like the option to be able to pass in an empty blacklist, or like in your example, to be able to merge different languages only if needed, not having them all forced by default.

As for scary algorithm - I don't think there's a way to properly optimize the search, given a bad word can be inside the generated hash.
This defeats all possible optimizations and will require a linear scan of the blacklist. That's one of the reasons I'd like the possibility to trash the blacklist - but still, for ease of use in most cases, if the user passes `null` for the blacklist, they'd get the default one, like in the previous version of the spec.

```ts
const sqids = new Sqids();                  // default blacklist
const sqids = new Sqids({ blacklist: [] }); // empty blacklist
```

**Reply by 4kimov** [reply]
Cool. I'm fine with the examples you gave, but what would it look like if a user had their own words they'd like blacklisted? (without providing an extra constructor param)

**Reply by tzvetkoff** [reply]
Perhaps something like
```ts
cosnt sqids = new Sqids({
  blacklist: [...defaultOptions.blacklist, customBlacklist]
});
```

Or maybe export the default blacklist as well as the `defaultOptions`?

**Reply by tzvetkoff** [reply]
We can even go further and have 1 default blacklist in the default package, and have another package with per-language blacklists, so users that need it can import the blacklist package and build their own set...
This will present another problem, though, with the evolution of said blacklists (although `decode` will still work.)

**Reply by 4kimov** [reply]
Yeah, this could work. I do enjoy packages that have auto-configured settings. Perhaps something like this is a good balance:

```js
const sqids = new Sqids(); // default blacklist
const sqids = new Sqids({ blacklist: [] }); // empty blacklist
const sqids = new Sqids({ blacklist: customBlacklist }); // custom blacklist
const sqids = new Sqids({ blacklist: [...defaultOptions.blacklist, ...customBlacklist] }); // default + custom blacklist
```

I'm just not sure all languages have the ability to handle something like this.

**Reply by tzvetkoff** [reply]
Perhaps not in 1 half-liner like JS/TS but it's still possible in a few lines.  That is, if you'd require special configuration, and the language built-in support is poor, the library can provide API for that - I'm using, for example, linked lists in C.

**Reply by 4kimov** [reply]
Cool, makes sense. [Adjusted spec](https://github.com/sqids/sqids-spec/blob/main/tests/blacklist.test.ts#L4-L35) for the above.

**Comment by 4kimov** [comment]
Any preference for passing the alphabet as a string or an array of chars to the constructor? [More discussion here](https://github.com/orgs/sqids/discussions/1).

**Reply by 4kimov** [reply]
Answered here: https://github.com/orgs/sqids/discussions/1#discussioncomment-6301537 (string) 👍
